compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py

```diff
@@ -224,10 +279,40 @@ def import_initializer(
             ).result
 
         self._nv_map[initializer_name] = converted_value
-        tensor_as_array = numpy_helper.to_array(initializer)
-        self.param_archive.add_buffer(actual_symbol_name, tensor_as_array)
+        if isinstance(initializer, onnx.TensorProto):
+            self.globals.append((initializer_name, actual_symbol_name))
         return converted_value
 
+    def save_params(self):
+        """
+        Only gets called if the arg `--save-params` is set to `True`.
+        Saving params requires iree-runtime, so putting the import here will avoid requiring it uniformly for the importer.
+        """
+        try:
+            import iree.runtime as rt
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                "iree-import-onnx requires iree runtime api for externalizing parameters. "
+                "For example: `pip install iree-base-runtime`"
+            ) from e
+        param_archive = rt.ParameterIndex()
+        param_count = 0
+        for name, actual_symbol_name in self.globals:
+            initializer = self._gi.initializer_map[name]
+            tensor_as_array = numpy_helper.to_array(
+                initializer, base_dir=self.param_data.data_dir
+            )
+            param_archive.add_buffer(actual_symbol_name, tensor_as_array)
+            param_count += 1
+            if (
+                self.param_data.initializer_threshold
+                and param_count % self.param_data.initializer_threshold == 0
+            ):
```

Comment:
Is the logic with `%` correct here? 

```python
    parser.add_argument(
        "--num-initializers-threshold",
        help="The maximum number of initializer tensors to be stored in-memory when creating a param archive.",
        type=int,
    )
```

> 2\. `--num-initializers-threshold` allows storing initializers to the irpa file in batches with a specified number of entries. This can reduce the memory overhead of first gathering all of the initializers, then saving them to the irpa at once.

The code doesn't seem to line up with the docs or PR description, unless I'm misreading it.

Reply:
Yeah, if the `param_count % initializer_threshold == 0`, then `param_count` is a multiple of the threshold, and we trigger saving the current param archive to the `.irpa` file. This will move in-memory tensors held in the param archive to the `.irpa` file.

Reply:
Oooooh I see. So like this?

```python
param_archive = rt.ParameterIndex()
for i in xrange(0, len(self.globals)):
  param_archive.add_buffer(...)

  # Observed 'threshold' parameters already, flush from memory to disk.
  if i % initializer_threshold == 0:
    param_archive = param_archive.create_archive_file(...)

# Flush any remaining parameters to disk.
param_archive = param_archive.create_archive_file(...)
```

If so, a few observations:

* Not all parameters are the same size, so a threshold based on the number of parameters could quite easily hit a degenerate case. Consider having 1000 parameters, with the first 10 being 1GB and the last 990 being 1KB. Flushing after even 2-3 would use substantial RAM at the start, then the flushing overhead would be slow for the long tail. Maybe threshold by memory used, not by parameter count?
* What does the data flow actually look like under these python API calls? The function names don't make it obvious to me what the costs are.
* This variable overwriting via `param_archive = param_archive.create_archive_file()` is sketchy on the surface... is the intent to append to a single file on disk? Or write to multiple files?

Reply:
Yeah, this is the result of some trial and error. I've basically run this on my laptop with another terminal checking memory usage and noticed flushing the parameters early takes more time but reduces the memory overhead. And I've definitely noticed that some batches are much larger than others, but I'm not sure how to set it up to flush at memory thresholds. I think maybe an good approach would be to count the number of elements in each param and multiply it by the bitwidth of the data type, then if adding a new param would exceed a bit threshold, then flush the archive before adding it. I'll try to mock this up in the next patch.

Calling `create_archive_file` seems to do the following (again, learned by trial and error):

1. params that already exist in the irpa file, and which correspond to references in the `param_archive`, do not get modified.
2. new params that exist in-memory get written to the irpa file.
3. setting `param_archive` equal to the result seems to convert the `param_archive` to only store references to the irpa file, so when we add more params to this, it will then contain both the existing irpa file references in addition to the newly added in-memory params.

Reply:
Let me know what you think of the changes!

Reply:
LGTM, except for

> Calling `create_archive_file` seems to do the following (again, learned by trial and error)

I'd rather make the APIs explicit. It seems like you're relying on undocumented side effects here, so I'm worried that we're in https://www.hyrumslaw.com/ territory :P

Reply:
Yeah, that's a good point. I'm doing a bit more digging into the API (it's calling this: <https://github.com/iree-org/iree/blob/cb593895d297ae2e79cef6fef0b7157096cb5b4a/runtime/bindings/python/io.cc#L424-L428>).

It might be better to hold two parameter indices: a target_index for the irpa file contents, and a running_index for in-memory params. I'll update when I know more